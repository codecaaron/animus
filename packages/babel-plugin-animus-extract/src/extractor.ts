import type { NodePath, types as BabelTypes } from '@babel/core';
import type { CallExpression } from '@babel/types';

import { generateCSS } from './css-generator';
import { createRuntimeReplacement } from './transformer';
import type { AnimusExtractState, ComponentData, ExtractedChainCall } from './types';
import { generateComponentId } from './utils';

export function isAnimusBuilder(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  // Direct call: animus.styles()
  if (
    t.isMemberExpression(callee) &&
    t.isIdentifier(callee.object) &&
    (callee.object.name === 'animus' || callee.object.name === state.animusBinding)
  ) {
    return true;
  }

  // Chained call: animus.styles().states()
  if (t.isMemberExpression(callee) && t.isCallExpression(callee.object)) {
    const objectPath = path.get('callee.object') as NodePath<CallExpression>;
    return isAnimusBuilder(objectPath, state);
  }

  return false;
}

function isAnimusChainCall(path: NodePath<CallExpression>, state: AnimusExtractState): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  return t.isMemberExpression(callee) &&
    (isAnimusBuilder(path, state) ||
     (t.isCallExpression(callee.object) && isAnimusChainCall(path.get('callee.object') as NodePath<CallExpression>, state)));
}

export function extractAnimusComponent(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): void {
  const chain = extractChain(path, state);
  if (chain.length === 0) return;

  const componentData = processChain(chain, state);

  // Generate CSS for this component
  const css = generateCSS(componentData, state);
  if (css) {
    state.cssOutput.set(componentData.id, css);
  }

  // Store component data for JSX transformation
  if (componentData.name) {
    state.animusComponents.set(componentData.name, componentData);
  }

  // Replace the original expression with runtime version
  const replacement = createRuntimeReplacement(componentData, state);
  path.replaceWith(replacement);
}

function extractChain(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): ExtractedChainCall[] {
  const chain: ExtractedChainCall[] = [];
  let current: NodePath<CallExpression> | null = path;

  while (current && isAnimusChainCall(current, state)) {
    const { types: t } = state;
    const {callee} = current.node;

    if (t.isMemberExpression(callee) && t.isIdentifier(callee.property)) {
      const methodName = callee.property.name;
      const args = current.node.arguments;

      chain.unshift({ methodName, args });

      if (t.isCallExpression(callee.object)) {
        current = current.get('callee.object') as NodePath<CallExpression>;
      } else {
        current = null;
      }
    } else {
      current = null;
    }
  }

  return chain;
}

function processChain(
  chain: ExtractedChainCall[],
  state: AnimusExtractState
): ComponentData {
  const { types: t } = state;
  const componentData: ComponentData = {
    id: generateComponentId(),
  };

  for (const { methodName, args } of chain) {
    switch (methodName) {
      case 'styles':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.styles = extractObjectLiteral(args[0], state);
        }
        break;

      case 'states':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.states = extractObjectLiteral(args[0], state);
        }
        break;

      case 'variants':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.variants = extractObjectLiteral(args[0], state);
        }
        break;

      case 'groups':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.groups = extractObjectLiteral(args[0], state);
        }
        break;

      case 'props':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.props = extractObjectLiteral(args[0], state);
        }
        break;

      case 'asElement':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.element = args[0].value;
        }
        break;

      case 'displayName':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.name = args[0].value;
        }
        break;
    }
  }

  return componentData;
}

function extractObjectLiteral(
  node: BabelTypes.ObjectExpression,
  state: AnimusExtractState
): Record<string, any> {
  const { types: t } = state;
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) || t.isObjectMethod(prop)) {
      const key = getPropertyKey(prop, state);
      if (key && t.isObjectProperty(prop)) {
        const value = extractValue(prop.value, state);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

function getPropertyKey(
  prop: BabelTypes.ObjectProperty | BabelTypes.ObjectMethod,
  state: AnimusExtractState
): string | null {
  const { types: t } = state;

  if (t.isIdentifier(prop.key)) {
    return prop.key.name;
  } else if (t.isStringLiteral(prop.key)) {
    return prop.key.value;
  }

  return null;
}

function extractValue(node: BabelTypes.Node, state: AnimusExtractState): any {
  const { types: t } = state;

  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node, state);
  } else if (t.isArrayExpression(node)) {
    return node.elements.map(el => el ? extractValue(el, state) : null);
  } else if (t.isUnaryExpression(node) && node.operator === '-' && t.isNumericLiteral(node.argument)) {
    return -node.argument.value;
  }

  // For dynamic values, return undefined to skip extraction
  return undefined;
}
