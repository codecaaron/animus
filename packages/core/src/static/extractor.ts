import * as parser from '@babel/parser';
// Handle the babel/traverse CommonJS export issue
// @ts-ignore - babel/traverse has complex module exports
import traverseDefault from '@babel/traverse';
import * as t from '@babel/types';

const traverse = (traverseDefault as any).default || traverseDefault;

import type { NodePath } from '@babel/traverse';

/**
 * Extracted style data from an Animus component
 */
export interface ExtractedStyles {
  componentName?: string;
  baseStyles?: Record<string, any>;
  variants?: Record<string, any> | Record<string, any>[];
  states?: Record<string, any>;
  groups?: string[];
  props?: Record<string, any>;
}

/**
 * Extract styles from Animus method chains in code
 */
export function extractStylesFromCode(code: string): ExtractedStyles[] {
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  });

  const extractedComponents: ExtractedStyles[] = [];

  traverse(ast as any, {
    CallExpression(path: NodePath<t.CallExpression>) {
      // Look for .styles() calls
      if (
        t.isMemberExpression(path.node.callee) &&
        t.isIdentifier(path.node.callee.property, { name: 'styles' })
      ) {
        const extracted: ExtractedStyles = {};

        // Try to get component name from variable declaration
        const varDeclarator = path.findParent((p: NodePath) =>
          p.isVariableDeclarator()
        );
        if (varDeclarator && t.isVariableDeclarator(varDeclarator.node)) {
          if (t.isIdentifier(varDeclarator.node.id)) {
            extracted.componentName = varDeclarator.node.id.name;
          }
        }

        // Extract styles object from first argument
        const stylesArg = path.node.arguments[0];
        if (t.isObjectExpression(stylesArg)) {
          extracted.baseStyles = extractObjectLiteral(stylesArg);
        }

        // Look for chained method calls by traversing up the AST
        // We need to find the full chain starting from .styles()
        const chainedCalls: any[] = [];
        let currentPath: any = path;

        // Traverse up to find all chained calls
        while (currentPath) {
          if (
            t.isCallExpression(currentPath.node) &&
            t.isMemberExpression(currentPath.node.callee) &&
            t.isIdentifier(currentPath.node.callee.property)
          ) {
            chainedCalls.push({
              method: currentPath.node.callee.property.name,
              args: currentPath.node.arguments,
            });
          }

          // Check if parent is also a call expression (chained)
          if (
            currentPath.parentPath &&
            t.isMemberExpression(currentPath.parentPath.node) &&
            currentPath.parentPath.parentPath &&
            t.isCallExpression(currentPath.parentPath.parentPath.node)
          ) {
            currentPath = currentPath.parentPath.parentPath;
          } else {
            break;
          }
        }

        // Process chained calls in order
        const variantCalls: any[] = [];

        for (const call of chainedCalls) {
          if (call.method === 'variant' && call.args[0]) {
            const variantArg = call.args[0];
            if (t.isObjectExpression(variantArg)) {
              variantCalls.push(extractObjectLiteral(variantArg));
            }
          } else if (call.method === 'states' && call.args[0]) {
            const statesArg = call.args[0];
            if (t.isObjectExpression(statesArg)) {
              extracted.states = extractObjectLiteral(statesArg);
            }
          } else if (call.method === 'groups' && call.args[0]) {
            const groupsArg = call.args[0];
            if (t.isObjectExpression(groupsArg)) {
              const groups = extractObjectLiteral(groupsArg);
              extracted.groups = Object.keys(groups).filter(
                (key) => groups[key]
              );
            }
          } else if (call.method === 'props' && call.args[0]) {
            const propsArg = call.args[0];
            if (t.isObjectExpression(propsArg)) {
              extracted.props = extractObjectLiteral(propsArg);
            }
          }
        }

        // Store variants - if multiple, keep as array; if single, flatten
        if (variantCalls.length > 1) {
          extracted.variants = variantCalls;
        } else if (variantCalls.length === 1) {
          extracted.variants = variantCalls[0];
        }

        extractedComponents.push(extracted);
      }
    },
  });

  return extractedComponents;
}

/**
 * Extract a literal object from an AST ObjectExpression
 */
function extractObjectLiteral(node: t.ObjectExpression): Record<string, any> {
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) && !prop.computed) {
      const key = t.isIdentifier(prop.key)
        ? prop.key.name
        : t.isStringLiteral(prop.key)
          ? prop.key.value
          : null;

      if (key) {
        const value = extractValue(prop.value);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

/**
 * Extract a literal value from an AST node
 */
function extractValue(node: t.Node): any {
  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isNullLiteral(node)) {
    return null;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node);
  } else if (t.isArrayExpression(node)) {
    return node.elements
      .filter((el): el is t.Expression | t.SpreadElement => el !== null)
      .map((el) => (t.isExpression(el) ? extractValue(el) : undefined))
      .filter((v) => v !== undefined);
  } else if (t.isTemplateLiteral(node) && node.expressions.length === 0) {
    // Simple template literal with no expressions
    return node.quasis[0].value.cooked;
  }

  // Return undefined for non-literal expressions
  return undefined;
}
