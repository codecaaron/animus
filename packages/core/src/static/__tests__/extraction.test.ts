import { extractStylesFromCode } from '../extractor';
import { CSSGenerator } from '../generator';
import { minimalSpace, testTheme } from './testConfig';

describe('Animus Static Extraction', () => {
  describe('Style Extraction', () => {
    it('extracts basic styles', () => {
      const code = `
        const Button = animus
          .styles({
            padding: 10,
            color: 'blue',
            fontSize: 16
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      expect(extracted[0]).toEqual({
        componentName: 'Button',
        baseStyles: {
          padding: 10,
          color: 'blue',
          fontSize: 16,
        },
      });
    });

    it('extracts pseudo-selectors', () => {
      const code = `
        const Link = animus
          .styles({
            color: 'blue',
            textDecoration: 'none',
            '&:hover': {
              color: 'darkblue',
              textDecoration: 'underline'
            },
            '&:active': {
              color: 'purple'
            }
          })
          .asElement('a');
      `;

      const extracted = extractStylesFromCode(code);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      expect(extracted[0]).toEqual({
        componentName: 'Link',
        baseStyles: {
          color: 'blue',
          textDecoration: 'none',
          '&:hover': {
            color: 'darkblue',
            textDecoration: 'underline',
          },
          '&:active': {
            color: 'purple',
          },
        },
      });
    });

    it('extracts responsive values', () => {
      const code = `
        const ResponsiveBox = animus
          .styles({
            padding: { _: 10, sm: 20, md: 30 },
            margin: [5, 10, 15, 20],
            fontSize: 16
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      expect(extracted[0]).toEqual({
        componentName: 'ResponsiveBox',
        baseStyles: {
          padding: { _: 10, sm: 20, md: 30 },
          margin: [5, 10, 15, 20],
          fontSize: 16,
        },
      });
    });

    it('extracts variants', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px',
            border: 'none'
          })
          .variant({
            prop: 'size',
            variants: {
              small: { padding: '4px 8px', fontSize: 14 },
              large: { padding: '12px 24px', fontSize: 18 }
            }
          })
          .variant({
            prop: 'color',
            variants: {
              primary: { backgroundColor: 'blue', color: 'white' },
              secondary: { backgroundColor: 'gray', color: 'black' }
            }
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      const buttonStyle = extracted[0] as any;

      expect(buttonStyle.componentName).toBe('Button');
      expect(buttonStyle.baseStyles).toEqual({
        padding: '8px 16px',
        border: 'none',
      });

      // Verify both variants are extracted
      expect(buttonStyle.variants).toHaveLength(2);

      // Check size variant
      expect(buttonStyle.variants[0]).toEqual({
        prop: 'size',
        variants: {
          small: { padding: '4px 8px', fontSize: 14 },
          large: { padding: '12px 24px', fontSize: 18 },
        },
      });

      // Check color variant
      expect(buttonStyle.variants[1]).toEqual({
        prop: 'color',
        variants: {
          primary: { backgroundColor: 'blue', color: 'white' },
          secondary: { backgroundColor: 'gray', color: 'black' },
        },
      });
    });

    it('extracts states', () => {
      const code = `
        const Input = animus
          .styles({
            padding: '8px',
            border: '1px solid gray'
          })
          .states({
            disabled: { opacity: 0.5, cursor: 'not-allowed' },
            error: { borderColor: 'red' },
            focus: { borderColor: 'blue', outline: 'none' }
          })
          .asElement('input');
      `;

      const extracted = extractStylesFromCode(code);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      expect(extracted[0]).toEqual({
        componentName: 'Input',
        baseStyles: {
          padding: '8px',
          border: '1px solid gray',
        },
        states: {
          disabled: { opacity: 0.5, cursor: 'not-allowed' },
          error: { borderColor: 'red' },
          focus: { borderColor: 'blue', outline: 'none' },
        },
      });
    });

    it('extracts groups and props', () => {
      const code = `
        const Box = animus
          .styles({
            display: 'flex'
          })
          .groups({
            space: true,
            color: true
          })
          .props({
            gap: { property: 'gap', scale: 'space' }
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      expect(extracted[0]).toEqual({
        componentName: 'Box',
        baseStyles: {
          display: 'flex',
        },
        groups: ['space', 'color'],
        props: {
          gap: { property: 'gap', scale: 'space' },
        },
      });
    });
  });

  describe('CSS Generation', () => {
    let generator: CSSGenerator;

    beforeEach(() => {
      generator = new CSSGenerator();
    });

    it('generates component CSS', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px',
            backgroundColor: 'blue',
            color: 'white',
            '&:hover': {
              backgroundColor: 'darkblue'
            }
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('component-css-generation');
    });

    it('generates CSS with variants', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px'
          })
          .variant({
            prop: 'size',
            variants: {
              small: { padding: '4px 8px' },
              large: { padding: '12px 24px' }
            }
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('variant-css-generation');
    });

    it('generates CSS with states', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px'
          })
          .states({
            disabled: { opacity: 0.5 },
            loading: { cursor: 'wait' }
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('state-css-generation');
    });

    it('generates atomic utilities for groups', () => {
      const code = `
        const Box = animus
          .styles({
            display: 'flex'
          })
          .groups({
            space: true
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const groupDefs = { space: minimalSpace };
      const result = generator.generateFromExtracted(
        extracted[0],
        groupDefs,
        testTheme
      );
      expect(result.css).toMatchSnapshot('atomic-utilities-generation');
    });

    it('demonstrates shorthand expansion in utilities', () => {
      const code = `
        const Container = animus
          .styles({
            width: '100%'
          })
          .groups({
            space: true
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      // Only include mx and py to show shorthand expansion
      const groupDefs = {
        space: {
          mx: minimalSpace.mx,
          py: minimalSpace.py,
        },
      };
      const result = generator.generateFromExtracted(
        extracted[0],
        groupDefs,
        testTheme
      );
      expect(result.css).toMatchSnapshot('shorthand-expansion-demo');
    });
  });
});
