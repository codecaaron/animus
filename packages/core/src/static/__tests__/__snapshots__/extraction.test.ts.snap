// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Animus Static Extraction CSS Generation demonstrates shorthand expansion in utilities: shorthand-expansion-demo 1`] = `
"/* Base Styles */
.animus-Container-c9r {
  width: 100%;
}

/* Atomic Utilities from Groups */
/* Group: space */"
`;

exports[`Animus Static Extraction CSS Generation generates CSS with states: state-css-generation 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 8px 16px;
}

/* State: disabled */
.animus-Button-b6n-state-disabled {
  opacity: 0.5;
}

/* State: loading */
.animus-Button-b6n-state-loading {
  cursor: wait;
}"
`;

exports[`Animus Static Extraction CSS Generation generates CSS with variants: variant-css-generation 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 8px 16px;
}

/* Variant: size="small" */
.animus-Button-b6n-size-small {
  padding: 4px 8px;
}

/* Variant: size="large" */
.animus-Button-b6n-size-large {
  padding: 12px 24px;
}"
`;

exports[`Animus Static Extraction CSS Generation generates atomic utilities for groups: atomic-utilities-generation 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  display: flex;
}

/* Atomic Utilities from Groups */
/* Group: space */"
`;

exports[`Animus Static Extraction CSS Generation generates component CSS: component-css-generation 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 8px 16px;
  background-color: blue;
  color: white;
}
.animus-Button-b6n:hover {
  background-color: darkblue;
}"
`;
